src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -2992,6 +2992,183 @@ public static Vector512<float> Invoke(Vector512<float> x)
+                }
+
+                Vector256<float> y = ux.AsSingle();
+                Vector256<float> z = ExpOperator.Invoke(y - Vector256.Create(LOGV).AsSingle());
+                return Vector256.Create(HALFV).AsSingle() * (z + Vector256.Create(INVV2).AsSingle() * 1f / z);
+            }
+
+#if NET8_0_OR_GREATER
+            public static Vector512<float> Invoke(Vector512<float> x)
+            {
+                Vector512<uint> ux = x.AsUInt32() & Vector512.Create(SIGN_MASK);
+                if (Vector512.GreaterThanAny(ux, Vector512.Create(ARG_MAX)))
+                {
+                    return Vector512.Create(Invoke(x.GetLower()), Invoke(x.GetUpper()));
+                }
+
+                Vector512<float> y = ux.AsSingle();
+                Vector512<float> z = ExpOperator.Invoke(y - Vector512.Create(LOGV).AsSingle());
+                return Vector512.Create(HALFV).AsSingle() * (z + Vector512.Create(INVV2).AsSingle() * 1f / z);
+            }
+#endif
+        }
+
+        /// <summary>MathF.Tanh(x)</summary>
+        private readonly struct TanhOperator : IUnaryOperator
+        {
+            // This code is based on `vrs4_tanhf` from amd/aocl-libm-ose
+            // Copyright (C) 2008-2022 Advanced Micro Devices, Inc. All rights reserved.
+            //
+            // Licensed under the BSD 3-Clause "New" or "Revised" License
+            // See THIRD-PARTY-NOTICES.TXT for the full license text
+
+            // To compute vrs4_tanhf(v_f32x4_t x)
+            // Let y = |x|
+            // If 0 <= y < 0x1.154246p3
+            //    Let z = e^(-2.0 * y) - 1      -(1)
+            //
+            //    Using (1), tanhf(y) can be calculated as,
+            //    tanhf(y) = -z / (z + 2.0)
+            //
+            // For other cases, call scalar tanhf()
+            //
+            // If x < 0, then we use the identity
+            //    tanhf(-x) = -tanhf(x)
+
+            private const uint V4_TANHF_ARG_MAX = 0x410AA123;
+            private const uint V4_TANHF_SIGN_MASK = 0x7FFFFFFF;
+
+            public static float Invoke(float x) => MathF.Tanh(x);
+
+            public static Vector128<float> Invoke(Vector128<float> x)
+            {
+                Vector128<uint> ux = x.AsUInt32();
+                Vector128<uint> sign = ux & Vector128.Create(~V4_TANHF_SIGN_MASK);
+
+                ux &= Vector128.Create(V4_TANHF_SIGN_MASK);
+                if (Vector128.GreaterThanAny(ux, Vector128.Create(V4_TANHF_ARG_MAX)))
+                {
+                    return Vector128.Create(
+                        MathF.Tanh(x.GetElement(0)),
+                        MathF.Tanh(x.GetElement(1)),
+                        MathF.Tanh(x.GetElement(2)),
+                        MathF.Tanh(x.GetElement(3)));
+                }
+
+                Vector128<float> y = ux.AsSingle();
+                Vector128<float> z = ExpOperator.Invoke(Vector128.Create(-2f) * y) - Vector128.Create(1f);
```

Comment:
The amd implementation does `-1f` but it has a comment saying it's doing `+1f`. I'm assuming the `-1f` is correct.

Reply:
-1 is correct.

The algorithm is $`{\frac {e^{2x} - 1}{e^{2x} + 1}}`$

It is then breaking this down into $`{\frac {-(e^{2|x|} - 1)}{({e^{2|x|} + 1) + 2}}}`$, which functionally computes `-tanh(x)`. It then toggles the sign of `x`, giving it the correct value.

Not quite sure why they did it this way since it could've been done the "normal" way without the need to toggle and without the need to `(z -1) + 2`. -- There's theoretically some very small rounding difference, but it doesn't look like something they're explicitly accounting for here.